spec/design_topics/data_dependent_output_shapes.md

```diff
@@ -0,0 +1,15 @@
+(data-dependent-output-shapes)=
+
+# Data-dependent output shapes
+
+Array libraries which build array computation graphs commonly employ static memory allocation techniques for performance optimization. In order to efficiently perform static memory allocation, such libraries must be able to infer array sizes during ahead-of-time (AOT) compilation. Functions and operations which are value-dependent present difficulties for such libraries, as array sizes cannot be inferred AOT without also knowing the contents of the respective arrays.
```

Comment:
I'm not sure I would describe what Dask uses known shapes for as "static memory allocation," though it does indeed have major performance implications nonetheless.

It's also a little confusing to use "ahead-of-time compilation" here because in JAX it actually arises for just-in-time compilation, too.

Static shapes are definitely relevant for performance, but they are also relevant just for having a sane data model, e.g., something you could imagine verifying with a static type checker. My suggested revision:

```suggestion
Array libraries which build array computation graphs commonly employ static analysis that relies upon known shapes. For example, JAX requires known array sizes when compiling code, in order to perform static memory allocation. Functions and operations which are value-dependent present difficulties for such libraries, as array sizes cannot be inferred AOT without also knowing the contents of the respective arrays.
```

Reply:
Thanks for the suggestion, @shoyer!